astropy/io/fits/fitsrec.py

```diff
@@ -1140,7 +1147,7 @@ def _scale_back(self, update_heap_pointers=True):
                     # The VLA has potentially been updated, so we need to
                     # update the array descriptors
                     raw_field[:] = 0  # reset
-                    npts = [len(arr) for arr in self._converted[name]]
+                    npts = [np.prod(arr.shape) for arr in self._converted[name]]
```

Comment:
Here is where the descriptor is recalculated just before to be written to the file.